x/wasm/internal/keeper/handler_plugin_encoders.go

```diff
@@ -0,0 +1,292 @@
+	}
+}
+
+func EncodeStargateMsg(unpacker codectypes.AnyUnpacker) StargateEncoder {
+	return func(sender sdk.AccAddress, msg *wasmvmtypes.StargateMsg) ([]sdk.Msg, error) {
+		any := codectypes.Any{
+			TypeUrl: msg.TypeURL,
+			Value:   msg.Value,
+		}
+		var sdkMsg sdk.Msg
+		if err := unpacker.UnpackAny(&any, &sdkMsg); err != nil {
+			return nil, sdkerrors.Wrap(types.ErrInvalidMsg, fmt.Sprintf("Cannot unpack proto message with type URL: %s", msg.TypeURL))
+		}
+		if err := codectypes.UnpackInterfaces(sdkMsg, unpacker); err != nil {
+			return nil, sdkerrors.Wrap(types.ErrInvalidMsg, fmt.Sprintf("UnpackInterfaces inside msg: %s", err))
+		}
+		return []sdk.Msg{sdkMsg}, nil
+	}
+}
+
+func EncodeWasmMsg(sender sdk.AccAddress, msg *wasmvmtypes.WasmMsg) ([]sdk.Msg, error) {
+	switch {
+	case msg.Execute != nil:
+		coins, err := convertWasmCoinsToSdkCoins(msg.Execute.Send)
+		if err != nil {
+			return nil, err
+		}
+
+		sdkMsg := types.MsgExecuteContract{
+			Sender:   sender.String(),
+			Contract: msg.Execute.ContractAddr,
+			Msg:      msg.Execute.Msg,
+			Funds:    coins,
+		}
+		return []sdk.Msg{&sdkMsg}, nil
+	case msg.Instantiate != nil:
+		coins, err := convertWasmCoinsToSdkCoins(msg.Instantiate.Send)
+		if err != nil {
+			return nil, err
+		}
+
+		sdkMsg := types.MsgInstantiateContract{
+			Sender:  sender.String(),
+			CodeID:  msg.Instantiate.CodeID,
+			Label:   msg.Instantiate.Label,
+			InitMsg: msg.Instantiate.Msg,
+			Funds:   coins,
+		}
+		return []sdk.Msg{&sdkMsg}, nil
+	case msg.Migrate != nil:
+		sdkMsg := types.MsgMigrateContract{
+			Sender:     sender.String(),
+			Contract:   msg.Migrate.ContractAddr,
+			CodeID:     msg.Migrate.NewCodeID,
+			MigrateMsg: msg.Migrate.Msg,
+		}
+		return []sdk.Msg{&sdkMsg}, nil
+	default:
+		return nil, sdkerrors.Wrap(types.ErrUnknownMsg, "unknown variant of Wasm")
+	}
+}
+
+func EncodeIBCMsg(ctx sdk.Context, sender sdk.AccAddress, contractIBCPortID string, msg *wasmvmtypes.IBCMsg) ([]sdk.Msg, error) {
+	switch {
+	case msg.CloseChannel != nil:
+		return []sdk.Msg{&channeltypes.MsgChannelCloseInit{
+			PortId:    PortIDForContract(sender),
+			ChannelId: msg.CloseChannel.ChannelID,
+			Signer:    sender.String(),
+		}}, nil
+	case msg.Transfer != nil:
+		amount, err := convertWasmCoinToSdkCoin(msg.Transfer.Amount)
+		if err != nil {
+			return nil, sdkerrors.Wrap(err, "amount")
+		}
+		portID := ibctransfertypes.ModuleName //todo: port can be customized in genesis. make this more flexible
+		msg := &ibctransfertypes.MsgTransfer{
+			SourcePort:       portID,
+			SourceChannel:    msg.Transfer.ChannelID,
+			Token:            amount,
+			Sender:           sender.String(),
+			Receiver:         msg.Transfer.ToAddress,
+			TimeoutHeight:    convertWasmIBCTimeoutHeightToCosmosHeight(msg.Transfer.TimeoutBlock),
+			TimeoutTimestamp: convertWasmIBCTimeoutTimestampToCosmosTimestamp(msg.Transfer.TimeoutTimestamp),
+		}
+		return []sdk.Msg{msg}, nil
+	default:
```

Comment:
Very cool, so the ones that can be easily translated into send messages are taken care of here.
And the one we cannot gets `ErrUnknownMsg`, so the IbcMsgHandler can take it on (and directly call the keeper)

Reply:
ja, anything that can not be translated into a sdk msg type can be handled by a new handler impl. No need to mess with the encoders as before. 😄